python/taichi/core/util.py

```diff
@@ -441,11 +439,46 @@ def require_version(major, minor=None, patch=None):
 
 at_startup()
 
-device_string = 'cpu only' if not tc_core.with_cuda() else 'cuda {}'.format(
-    tc_core.cuda_version())
-print(
-    f'[Taichi] version {tc_core.get_version_string()}, {device_string}, commit {tc_core.get_commit_hash()[:8]}, python {sys.version_info[0]}.{sys.version_info[1]}.{sys.version_info[2]}'
-)
+
+def _print_taichi_header():
+    dev_mode = not is_release()
+
+    header = '[Taichi] '
+    if dev_mode:
+        header += '<dev mode>, '
+    else:
+        header += f'version {tc_core.get_version_string()}, '
+
+    supported_archs = ['cpu']
+    if tc_core.with_cuda():
+        supported_archs.append('cuda')
+    if tc_core.with_opengl():
+        supported_archs.append('opengl')
+    if tc_core.with_metal():
+        supported_archs.append('metal')
+    if len(supported_archs) == 1:
+        supported_archs[0] = 'cpu only'
+    archs_str = ', '.join(sorted(supported_archs))
+    header += f'supported archs: [{archs_str}], '
+
+    commit_hash = tc_core.get_commit_hash()
+    if dev_mode:
+        repo = get_repo()
+        try:
+            commit_hash = str(repo.head.commit)
+        except:
+            pass
```

Comment:
Thanks for fixing everything here!

Actually `tc_core.get_commit_hash()` gives you the commit the last time `cmake` was executed instead of latest nightly release commit: https://github.com/taichi-dev/taichi/blob/213a3c7c02823a95d67ec2c9cf562c7f0389746c/CMakeLists.txt#L18-L22 

These info will be written into `taichi/common/version.h`
https://github.com/taichi-dev/taichi/blob/213a3c7c02823a95d67ec2c9cf562c7f0389746c/CMakeLists.txt#L65-L73

I think `ti` should show the latest **compiled** commit instead of **HEAD**, since it might be the case one developer just checks out the commit without compilation, then the git commit info displayed here can be a bit misleading. 

How about using a [pre-build step](https://stackoverflow.com/questions/18427877/add-custom-build-step-in-cmake) to always update `taichi/common/version.h` before building, so that the git commit hash in the compiled version always reflect the actual commit being built?

(Of course, if someone's local changes are not committed, then the changes to be committed will not be reflected, unless we add an extra `bool has_changes_to_commit = true/false` to `taichi/common/version.h`...)

Reply:
OK, I will try this in the next few days. TBH i'm very rusty with CMake..

Reply:
I figured out how to do this one. I'll implement it now and have this PR merged :-)

Reply:
Thanks!!

Reply:
Done! Please take a look - I'll merge once CI passes if everything is good.

Reply:
Looks great! (weird that I'm making such a comment where I initiated the PR...)